File: android/app/src/main/java/au/com/dangarmarine/jacl/JaclAudio.kt
package au.com.dangarmarine.jacl

import android.media.MediaDataSource
import android.media.MediaPlayer
import android.util.Log

/**
 * Plays the game's sound channels. The interpreter (via RemGlk's rgschan.c)
 * sends play/stop/volume ops keyed by channel id; we resolve each sound number
 * to its blorb bytes ([sound]) and play it on a per-channel MediaPlayer.
 *
 * Blorb sounds are Ogg Vorbis / WAV (Android decodes these natively) fed
 * straight from memory via a MediaDataSource -- no temp files.
 */
class JaclAudio(private val sound: (Int) -> ByteArray?) {

    private val players = HashMap<Int, MediaPlayer>()
    /** Last requested volume per channel (0..1), applied once prepared. */
    private val volumes = HashMap<Int, Float>()
    /** When true (the Sound setting is off), play ops are ignored. */
    var muted = false
        set(value) {
            field = value
            if (value) releaseAll()
        }

    /** Play sound [snd] on [chan]; loops if [repeats] is -1. [vol] is the Glk
     *  0..0x10000 channel volume. Replaces whatever was on the channel. */
    fun play(chan: Int, snd: Int, repeats: Int, vol: Int) {
        stop(chan)
        if (muted) return
        val bytes = sound(snd) ?: return
        val v = (vol / 65536f).coerceIn(0f, 1f)
        volumes[chan] = v
        try {
            val mp = MediaPlayer()
            mp.setDataSource(ByteArrayMediaDataSource(bytes))
            mp.isLooping = (repeats == -1)
            mp.setOnPreparedListener {
                it.setVolume(v, v)
                it.start()
            }
            mp.setOnCompletionListener {
                if (!it.isLooping) { it.release(); if (players[chan] === it) players.remove(chan) }
            }
            mp.setOnErrorListener { p, _, _ -> p.release(); if (players[chan] === p) players.remove(chan); true }
            players[chan] = mp
            mp.prepareAsync()
        } catch (e: Exception) {
            Log.e("JACL", "sound play failed", e)
        }
    }

    fun stop(chan: Int) {
        players.remove(chan)?.let {
            try { it.reset() } catch (_: Exception) {}
            it.release()
        }
    }

    fun setVolume(chan: Int, vol: Int) {
        val v = (vol / 65536f).coerceIn(0f, 1f)
        volumes[chan] = v
        try { players[chan]?.setVolume(v, v) } catch (_: Exception) {}
    }

    /** Release every channel (leaving a game / Restart). */
    fun releaseAll() {
        players.values.forEach { try { it.reset(); it.release() } catch (_: Exception) {} }
        players.clear()
        volumes.clear()
    }
}

/** Feeds an in-memory byte array to MediaPlayer (API 23+). */
private class ByteArrayMediaDataSource(private val data: ByteArray) : MediaDataSource() {
    override fun readAt(position: Long, buffer: ByteArray, offset: Int, size: Int): Int {
        if (position >= data.size) return -1
        val n = minOf(size, data.size - position.toInt())
        System.arraycopy(data, position.toInt(), buffer, offset, n)
        return n
    }
    override fun getSize(): Long = data.size.toLong()
    override fun close() {}
}

File: ios/JACL/JaclAudio.swift
//  JaclAudio.swift
//  Plays the game's Glk sound channels.
//
//  The interpreter (via RemGlk's rgschan.c) emits play/stop/volume ops keyed by
//  channel id; we resolve each sound number to its blorb bytes, decode the Ogg
//  Vorbis to WAV (jacl_ogg_to_wav, AVFoundation can't play Ogg), and play it on
//  a per-channel AVAudioPlayer.

import Foundation
import AVFoundation

final class JaclAudio {
    /// Resolves a sound resource number to its raw blorb bytes (Ogg Vorbis).
    private let sound: (Int) -> Data?
    private var players: [Int: AVAudioPlayer] = [:]
    private var volumes: [Int: Float] = [:]

    /// When true (the Sound setting is off), play ops are ignored.
    var muted = false {
        didSet { if muted { stopAll() } }
    }

    init(sound: @escaping (Int) -> Data?) {
        self.sound = sound
        // Mix with other audio and respect the silent switch (ambient).
        try? AVAudioSession.sharedInstance().setCategory(.ambient, options: [.mixWithOthers])
        try? AVAudioSession.sharedInstance().setActive(true)
    }

    /// Play sound `snd` on `chan`; loops if `repeats` is -1. `vol` is the Glk
    /// 0..0x10000 channel volume. Replaces whatever was on the channel.
    func play(chan: Int, snd: Int, repeats: Int, vol: Int) {
        stop(chan: chan)
        guard !muted, let ogg = sound(snd), let wav = JaclAudio.oggToWav(ogg) else { return }
        let v = max(0, min(1, Float(vol) / 65536))
        volumes[chan] = v
        do {
            let player = try AVAudioPlayer(data: wav)
            player.volume = v
            player.numberOfLoops = (repeats == -1) ? -1 : 0
            player.prepareToPlay()
            player.play()
            players[chan] = player
        } catch {
            NSLog("JACL sound: %@", String(describing: error))
        }
    }

    func stop(chan: Int) {
        players[chan]?.stop()
        players[chan] = nil
    }

    func setVolume(chan: Int, vol: Int) {
        let v = max(0, min(1, Float(vol) / 65536))
        volumes[chan] = v
        players[chan]?.volume = v
    }

    /// Stop every channel (leaving a game / Restart / muting).
    func stopAll() {
        players.values.forEach { $0.stop() }
        players.removeAll()
    }

    /// Decode Ogg Vorbis bytes to WAV bytes AVAudioPlayer can take, via the C
    /// stb_vorbis helper. nil if the bytes aren't decodable.
    private static func oggToWav(_ ogg: Data) -> Data? {
        ogg.withUnsafeBytes { raw -> Data? in
            guard let base = raw.baseAddress else { return nil }
            var len: Int32 = 0
            guard let wav = jacl_ogg_to_wav(base, Int32(ogg.count), &len), len > 0 else { return nil }
            let data = Data(bytes: wav, count: Int(len))
            free(wav)
            return data
        }
    }
}
